### ISSUE TITLE
Improve inference speed

### ISSUE THREAD

**Issue by LeDuySon** [pytorch/serve#1054]
Why when i increase initial worker the inference times still unchanged? Can someone explain

**Comment by msaroufim** [comment]
Could you share a bit more about how you're measuring inference time?

One guess is that scaling workers should have no impact on a single inference but instead increase your total throughput by allowing more concurrent inferences

**Comment by LeDuySon** [comment]
Im writing a script to get handle time from model_metrics.log file. And making concurrent inferences about 133 requests with aiohttp, batch size 64

**Comment by msaroufim** [comment]
Can you share a small reproducer? `aiohttp` seems like what we should be using for all our examples instead of `requests` 

Also are you running on CPU/GPU? Can you share nvidia-smi or logs confirming number of workers are allocated

Can you try doubling the number of concurrent inferences and verify that the pattern persists

**Comment by LeDuySon** [comment]
Im running on CPU
Logs confirming number of workers:
![Screenshot from 2021-05-08 01-04-05](https://user-images.githubusercontent.com/33374938/117490945-b830e000-af99-11eb-8b08-9c727ddd4394.png)
I have tried about 1000 requests but nothing changed 
Im using roberta for sentiment analysis
Here is output from torchserve --start
![Screenshot from 2021-05-08 01-07-46](https://user-images.githubusercontent.com/33374938/117491228-1d84d100-af9a-11eb-9319-a2bcf21ea7c1.png)


Thanks for your reply!!!

**Comment by msaroufim** [comment]
Apologies for the back and forth but I still need a bit more detail

Can you send me logs and code I can run? You can paste in a gist if it's too long https://gist.github.com/

Including the `aiohttp` script, the `serve` logs, the `model_metrics.txt` and anything else you did - the more detail the better. I'd like to run the same thing you did to better understand the issue

**Comment by LeDuySon** [comment]
My mar file in here https://drive.google.com/file/d/1KL_I0zHAlzGL2uNjeQNkLYhHyBiGA0eG/view?usp=sharing 
Asynchronous ìnference requests with aiohttp 

`import requests
import json
import asyncio
import logging
import re
import sys
from typing import IO
import urllib.error
import urllib.parse
import time
import aiofiles
import aiohttp
from aiohttp import ClientSession
import pandas as pd

logger = logging.getLogger(__name__)

async def get_prediction(url: str, session: ClientSession, text: str) -> str:
    resp = await session.request(method="POST", url=url, data=text)
    resp.raise_for_status()
    logger.info("Got response [%s] for URL: %s", resp.status, url)
    result = await resp.text()
    return (text, result)

async def parse(url: str, session: ClientSession, text) -> set:
    predict = ""
    try:
        print("Process: ", text)
        predict = await get_prediction(url=url, session=session, text=text)
    except (
        aiohttp.ClientError,
        aiohttp.http_exceptions.HttpProcessingError,
    ) as e:
        logger.error(
            "aiohttp exception for %s [%s]: %s",
            url,
            getattr(e, "status", None),
            getattr(e, "message", None),
        )
        return predict
    except Exception as e:
        logger.exception(
            "Non-aiohttp exception occured:  %s", getattr(e, "__dict__", {})
        )
    return predict

async def write_database(file: IO, url: str, **kwargs) -> None:
    # res = await parse(url=url, **kwargs)
    # if not res:
    #     return None
    # async with aiofiles.open(file, "a") as f:
    #     for p in res:
    #         await f.write(f"{url}\t{p}\n")
    #     logger.info("Wrote results for source URL: %s", url)
    pass

async def predict_batch(url: str, texts: set) -> None:
    async with ClientSession() as session:
        tasks = []
        for text in texts:
            tasks.append(
                parse(url=url, session=session, text=text)
            )
        predictions = await asyncio.gather(*tasks)
        return predictions

async def predicts(url: str, texts: set) -> list:
    pass
    
    
def main(config, texts):
    req_url = f'http://localhost:8080/predictions/{config["model_name"]}'
    
    # if(len(texts) > config["batch_size"]):
    #     texts = texts[:config["batch_size"]]
        
    predictions = asyncio.run(predict_batch(url=req_url, texts=texts))
    with open("results.txt", "a") as f:
        for k, v in predictions:
            f.write(k + " " + v + "\n")`

**Comment by msaroufim** [comment]
Hi @LeDuySon sorry for the delay

Can you add `torch.set_num_threads(1)` at the top of your handler file? This substantially improved CPU performance for me.

I'm working on adding support for it in the base handler but just thought I'd update you in case you'd like to try it out first

**Comment by ricardomlyk** [comment]
Hi @msaroufim

`torch.set_num_threads(1)` really helps a lot for my resnet34 image classify implement which has very simple preprocess(resize and rgb2gray) and no postprocess.
Before i limit pytorch with `torch.set_num_threads(1)`, the CPU Loads soared up to 90% with a 64 cores CPU when model do prediction, and after i limit pytorch with `torch.set_num_threads(1)`, the CPU Loads kept at 20%.
And The GPU Loads from 1% to 80%.
What 's more, the QPS of model inceased from 16 to 600.

At first I thought this was a common problem, but when I tried to port a more complex model to torchserve, I didn’t find any abnormalities in QPS. Whether I set `torch.set_num_threads(1)` or not, QPS is the same.

Hope this would help you, and i am curious why this happens and why `torch.set_num_threads(1)` works sometimes
